Forward Error Correction

autumn-traveller opened the pull request:
# Overview
This PR adds support for forward error correction. Firstly, an abstract class for FEC schemes of any kind is added ("FEC Transformer"), which can be used to encode individual files and source blocks. Secondly, an instance of the class for a raptor-based FEC scheme is added. The mechanism used is very similar to the Raptor FEC encoding scheme but differs in the generation of random numbers for selecting symbol degrees, and in not generating Half symbols. The scheme used only generates LDPC symbols during the intermediate symbol generation.

Full list of Changes:

- Abstract class for FEC schemes
- New github workflow to test the build process on various ubuntu versions
- Small improvements in example transmitter
- Raptor-based FEC library as a git submodule
- FEC build option
- Fix bug where the default transport session id in the transmitter is different than the default in the receiver
- Re-order the IPsec struct to prevent an error in the build process
- Update the call to the SSL library to the latest version, to prevent a warning and to compy with future SSL versions

This PR (unintentionally) addresses Issue #1 

# Building with FEC

The library used for FEC is in a git submodule in another repository. To pull and build the raptor changes you need to pull the other submodule and then rebuild. Pull using `git submodule update --init` if you are in a previously checked out repsitory.

When building, raptor is enabled by default (but can be turned off) so build as usual.

# Testing the changes

To test that everything works you need an environment with packet loss. One way to do this is to modify the loopback device so that it experiences packet loss. There is a script included - `setup_packet_loss_on_loopback` which uses the `tc` command to adjust the loopback device to experience packet loss, as well as adding a routing entry for the default mulitcast address used by libflute to make it go via the lo device. After testin you can rever the changes with the `setup_packet_loss_on_loopback` script.

autumn-traveller committed:
reorder ipsec struct to fix compilation issues, change TSI in flute-transmitter to match receiver

autumn-traveller committed:
use switch cases for FEC type instead of if statements and added placeholders for future integration of Raptor FEC

autumn-traveller committed:
added FEC scheme as cli argument for transmitter and receiver demos

SteezyE committed:
added build script

SteezyE committed:
simplified installation guide

autumn-traveller committed:
fixed md5 calculation to no longer emit deprecated openssl warning

SteezyE committed:
added freeRaptor submodule

SteezyE committed:
modifying cmake and adding raptor fec files

autumn-traveller committed:
document and cleanup calculate_md5() function

autumn-traveller committed:
updated readme and .gitmodules for raptor submodule

SteezyE committed:
working on compiling/linking

autumn-traveller committed:
point submodule towards free raptor dev branch

SteezyE committed:
fixed linking/compiling using build script

autumn-traveller committed:
cmakelists integration of freeRaptor

autumn-traveller committed:
make raptor compiliation optional and define a RAPTOR_ENABLED macro

SteezyE committed:
add raptor submodule

SteezyE committed:
add remote raptor repo to .gitmodules

SteezyE committed:
changed submodule to fork of raptor lib

autumn-traveller committed:
write donwloaded files to current directory, when building build in debug mode by default, small fixes in readme,cmakelists,gitignore

autumn-traveller committed:
flute-receiver: add cli argument for destination directory for downloaded files

autumn-traveller committed:
handle invalid file paths properly, increment the correct ptr in encoding symbols to_payload() functions

autumn-traveller committed:
abstract class definition for FEC object coding

autumn-traveller committed:
fixed abstract definition so it compiles, and reintroduced FecOti struct since its too difficult to directly replace it

autumn-traveller committed:
integreate FecTransformer structure into fdt and file classes

autumn-traveller committed:
generation and parseing for raptor OTI, set codepoint in alc packets based on fec scheme in use

Etua committed:
Switch freeRaptor to raptor

autumn-traveller committed:
use 256kB for target sub block size instead of 1MB

Etua committed:
Added initial encoder implementation

SteezyE committed:
resolve merge conflict

SteezyE committed:
fix merge conflict

SteezyE committed:
resolve merge conflict

SteezyE committed:
remove old freeRaptor header

SteezyE committed:
cmake fix

Etua committed:
Check for nullptr in create_blocks

Etua committed:
Calculate bytes_read correctly

Etua committed:
Replace TODOs

Etua committed:
Merge remote-tracking branch 'origin/raptor_integration' into raptor_integration

autumn-traveller committed:
update submodule

autumn-traveller committed:
dont calculate raptor parameters in partitioning() function, hardcoded W = 16 MB to keep N = 1

autumn-traveller committed:
finished raptor encoding + sending, removed compact no code subclass and file, cleaned up cmakelistst for raptor build case

SteezyE committed:
integrated raptor decoding functionality with libflute

autumn-traveller committed:
finish decoder, start SBN and ESI at 0, added process_symbol() function for all FEC transformers

autumn-traveller committed:
fixes for running receiver

autumn-traveller committed:
update raptor test.c

autumn-traveller committed:
fixed bad return in check_source_completion and allow fec transformres to allocate the file buffer themselves

autumn-traveller committed:
simulate 1pcent packet loss during reception, but never on the FDT

autumn-traveller committed:
added github action to test the build process

autumn-traveller committed:
use matrix of build and os options + fix compilation when raptor is disabled or ssl doesnt define ex2 on ubu20

autumn-traveller committed:
remove vim swap files

autumn-traveller committed:
clean up build workflow

autumn-traveller committed:
option stop receiving after n files + minor fixes

SteezyE committed:
fixed double free error

autumn-traveller committed:
use mmap to read in files, dont simulate packet loss if there is no active FEC scheme

Etua committed:
Fix error with accessing FDT under simulated packet loss

Etua committed:
Free the Symbol data

Etua committed:
Get rid of code-simulated paket loss

autumn-traveller committed:
scripts for adding packet loss to loopback interface for testing

autumn-traveller committed:
deal with the last block case

autumn-traveller committed:
calculate buffer offset correctly for multi block files + use delete[] instead of delete

autumn-traveller committed:
only extract finished source blocks at the end of transmission

autumn-traveller committed:
small fixes: memory deallocation in raptor and derived class destructor calls in FEC + updated readme

dsilhavy commented on src/AlcPacket.cpp:
```diff
@@ -79,10 +80,16 @@ LibFlute::AlcPacket::AlcPacket(char* data, size_t len)
+      _fec_oti.encoding_id = FecScheme::Raptor;
+      break;
+    default:
+      throw "Only Compact No-Code FEC is supported";
```
Should we adjust this error message?

autumn-traveller replied:
Yes we should! I have corrected it in the latest commit

dsilhavy commented on src/AlcPacket.cpp:
```diff
@@ -110,19 +117,29 @@ LibFlute::AlcPacket::AlcPacket(char* data, size_t len)
+                          hdr_ptr += 4;
+                          break;
+                        case FecScheme::Raptor:
+                          //TODO
```
What is missing here?

autumn-traveller replied:
So the FLUTE specification requires the receiver be able to process FEC parameters from **both** the FDT and the EXT_FTI header extension, whichever one the transmitter chooses to use. However our implementation only ever uses the FDT to communicate these parameters, so we didnt implement the use of EXT_FTI extension.

autumn-traveller committed:
Small fixes + delete build.sh

Remove an unnecessary #ifdef in Receiver.h, and fix error messages
in AlcPacket.cpp

dsilhavy commented:
@autumn-traveller : @kuehnhammer is currently working on the FEC integration and harmonizing our two implementations (one of these is yours). Then we should be ready to merge.
